precompile/contracts/warp/contract_warp_handler.go

```diff
@@ -61,7 +61,8 @@ func handleWarpMessage(accessibleState contract.AccessibleState, input []byte, s
 	warpIndex := int(warpIndexInput) // This conversion is safe even if int is 32 bits because we checked above.
 	state := accessibleState.GetStateDB()
 	predicateBytes, exists := state.GetPredicateStorageSlots(ContractAddress, warpIndex)
-	predicateResultsBytes := accessibleState.GetBlockContext().GetPredicateResultsBytes()
+	extra := accessibleState.GetBlockContext().Extra()
+	predicateResultsBytes := predicate.GetPredicateResultBytes(extra)
```

Comment:
IMHO:
warp contract should not need to deal with parsing predicate results (we already parse them for verification). I'm looking at the upstream and seems they also don't include header in the block context. If we can be flexible in BlockContext (we should be), we don't actually need to abstract everything. I do think BlockContext should be the place where we can have custom stuff derived from block/header to present for the execution. 
warp contract should not know about extra at all
Predicaters does not need to be just for warp messaging, any other predicater should not re-implement all this parse-from-extra logic.
These are my observations after looking at just diffs, block context and warp handler. I haven't considered libevm or any other cyclic dependency. LMK if there is another concern that outweighs these considerations (especially parsing).

Reply:
This approach https://github.com/ava-labs/coreth/pull/679 tries to follow the comment above, let me know your thoughts.